### PR TITLE
DAOS-3317 control: add system stop subcommand to daos_shell

### PR DESCRIPTION
**Pull request by tanabarr**
"daos_shell -l <hostlist> system stop" will trigger controlled shutdown
(request will be sent to first element of hostlist assumed to be the
access point). "system member-query" will return the list of IO-servers
(members) joined to DAOS system.

* Create subcommands and gRPC plumbing for requests. Some functionality
  will be missing and implemented in future.

* Create types and functionality for SystemMembers and Membership

* DAOS-3316 control: send grpc stop to system membership.
  
  Send SystemStop requests (currently handler will kill rank over dRPC)
  from management service leader to entire system membership. Requests
  sent by MgmtSvcClient in parallel over gRPC. Requires storing dest
  address for member as peer-ip:grpc-listener-port for stop grpc reqs.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/mgmt_svc.go [hunk 1]
```diff
@@ -212,9 +212,9 @@ func (svc *mgmtSvc) Join(ctx context.Context, req *mgmtpb.JoinReq) (*mgmtpb.Join
 		svc.members = append(svc.members, &SystemMember{
 			Addr: req.GetAddr(), Uuid: req.GetUuid(), Rank: resp.GetRank(),
 		})
-		svc.mutex.Unlock()
 
 		svc.log.Debugf("MgmtSvc.members: %+v\n", svc.members)
+		svc.mutex.Unlock()
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: this was requested by @mjmac , I will make it consistent

[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/mgmt_svc.go]: To be clear, the lock isn't to protect the logger, it's to protect svc.members from unsynchronized reads. Rather than sprinkling locks around everywhere, though, I'd recommend creating a new type to wrap the membership slice (or map, so that you can approximate a set) with synchronized access methods.

[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: ok, will do that

File: src/control/client/system.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import (
+	"golang.org/x/net/context"
+
+	"github.com/daos-stack/daos/src/control/common"
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/pkg/errors"
+)
+
+// SystemStop will create a DAOS pool using provided parameters and return
+// uuid, list of service replicas and error (including any DER code from DAOS).
```
Comment (kjacque) [inline src/control/client/system.go]: Incorrect comment

File: src/control/client/system.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	if rpcResp.GetStatus() != 0 {
+		return nil, errors.Errorf("DAOS returned error code: %d\n",
+			rpcResp.GetStatus())
+	}
+
+	return common.MembersFromPB(rpcResp.Members), nil
+}
+
+// SystemQuery will create a DAOS pool using provided parameters and return
+// uuid, list of service replicas and error (including any DER code from DAOS).
```
Comment (kjacque) [inline src/control/client/system.go]: Incorrect comment

**Comment by kjacque** [comment]
I would still be in favor of splitting out the dRPC client changes if it's not an enormous pain to do so.

File: src/control/client/system.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+// Isolate protobuf encapsulation in client and don't expose to calling code.
+func (c *connList) SystemStop() ([]*common.SystemMember, error) {
+	mc, err := chooseServiceLeader(c.controllers)
+	if err != nil {
+		return nil, err
+	}
+
+	rpcReq := &mgmtpb.SystemStopReq{}
+
+	c.log.Debugf("DAOS system shutdown request: %s\n", rpcReq)
```
Comment (dpquigl) [inline src/control/client/system.go]: What does this look like as a string? Should it be a %v instead?

File: src/control/client/system.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	rpcResp, err := mc.getSvcClient().SystemStop(context.Background(), rpcReq)
+	if err != nil {
+		return nil, err
+	}
+
+	c.log.Debugf("DAOS system shutdown response: %s\n", rpcResp)
```
Comment (dpquigl) [inline src/control/client/system.go]: Id imagine the question above pertains to this as well.

File: src/control/cmd/dmg/system.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+package main
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// systemCmd is the struct representing the top-level system subcommand.
+type SystemCmd struct {
```
Comment (dpquigl) [inline src/control/cmd/dmg/system.go]: SystemCmd.SystemQuery reads like it stutters. Maybe just make the internals the command without System infront?

File: src/control/server/drpc.go [hunk 7]
```diff
@@ -162,15 +162,17 @@ func newDrpcCall(module int32, method int32, bodyMessage proto.Message) (*drpc.C
...
+func makeDrpcCall(client drpc.DomainSocketClient, module int32, method int32,
 	body proto.Message) (drpcResp *drpc.Response, err error) {
 
 	drpcCall, err := newDrpcCall(module, method, body)
 	if err != nil {
 		return drpcResp, errors.Wrap(err, "build drpc call")
 	}
 
+	client.Lock()
+	defer client.Unlock()
```
Comment (dpquigl) [inline src/control/server/drpc.go]: is this recommended practice? I understand you'd have to unlock in 4 places if you didn't do this but I always worry about relying on someone else to implicitly unlock a mutex. If this is how they recommend doing it I guess its ok.

File: src/control/server/harness.go [hunk 8]
```diff
@@ -104,6 +104,25 @@ func (h *IOServerHarness) GetManagementInstance() (*IOServerInstance, error) {
 	return h.instances[defaultManagementInstance], nil
 }
 
+// GetMSLeaderInstance returns a managed IO Server instance to be used as a
+// management target and fails if selected instance is not MS Leader.
+func (h *IOServerHarness) GetMSLeaderInstance() (*IOServerInstance, error) {
+	h.RLock()
```
Comment (dpquigl) [inline src/control/server/harness.go]: Is there a reason why you place the rwlock first in this struct but the mutex at the bottom in the client connection. If you put the client connection one up top could you just do h.Lock and h.Unlock on the mutex there as well?

File: src/control/client/system.go (hunk 4 above)
Comment (tanabarr) [reply src/control/client/system.go]: the proto request and response messages have Stringer interface implementations e.g. https://github.com/daos-stack/daos/pull/1150/files#diff-f92c50fadb4724c224b17824268423f5R85

File: src/control/client/system.go (hunk 5 above)
Comment (tanabarr) [reply src/control/client/system.go]: as above

File: src/control/server/drpc.go (hunk 7 above)
Comment (tanabarr) [reply src/control/server/drpc.go]: I'm not sure I understand the concern, the mutex on the socket client will be unlocked when makeDrpcCall returns. we don't really mean to hold the lock during checkDrpcResponse so I can unlock before then if we think that is significant.

File: src/control/server/harness.go (hunk 8 above)
Comment (tanabarr) [reply src/control/server/harness.go]: yes good point, have changed it to use an embedded type as you suggest

File: src/control/cmd/dmg/system.go (hunk 6 above)
Comment (tanabarr) [reply src/control/cmd/dmg/system.go]: done

File: src/control/client/system.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// KillRank Will terminate server running at given rank on pool specified by
+// uuid. Request will only be issued to a single access point.
+func (c *connList) KillRank(uuid string, rank uint32) ResultMap {
+	results := make(ResultMap)
+
+	mc, err := chooseServiceLeader(c.controllers)
+	if err != nil {
+		result := ClientResult{mc.getAddress(), nil, err}
```
Comment (kjacque) [inline src/control/client/system.go]: Guaranteed to always get an instance back even if there's an error? If so, I think it is better to say:
```
var resp *DaosResp
if err == nil {
	resp, err = mc.getSvcClient().KillRank(context.Background(),
		&mgmtpb.DaosRank{PoolUuid: uuid, Rank: rank})
}

result := ...
```

Comment (tanabarr) [reply src/control/client/system.go]: I wasn't spending too much time on this because will go away shortly but I will fix up anyway

Comment (tanabarr) [reply src/control/client/system.go]: nice to see we are on the same page because that is what I was contemplating

File: src/control/cmd/dmg/system.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+// Execute is run when systemStopCmd activates
+func (cmd *systemStopCmd) Execute(args []string) error {
+	msg := "SUCCEEDED: "
+
+	members, err := cmd.conns.SystemStop()
+	if err != nil {
+		msg = errors.WithMessagef(err, "FAILED").Error()
+	}
+	if len(members) > 0 {
+		msg += fmt.Sprintf(": %+v", members)
```
Comment (mjmac) [inline src/control/cmd/dmg/system.go]: This seems like it could be excessively verbose. For tiny clusters it might be fine, but for clusters with tens, hundreds, thousands of members, this won't scale well. Perhaps it'd be better to just log the number of members that succeeded?

Also, one pattern to think about with these kinds of commands is that we'll probably want to display more than a binary choice between "it 100% worked" and "at least one failed, so the entire thing looks like it failed".

e.g. 
SUCCEEDED: 62
FAILED: 2

or better, but we don't yet have hostlist support:
SUCCEEDED: fnord[00-23,25-55,57-63]
FAILED: fnord[24,56]

File: src/control/client/connect.go [hunk 11]
```diff
@@ -34,6 +34,7 @@ import (
 	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
 	"github.com/daos-stack/daos/src/control/logging"
 	"github.com/daos-stack/daos/src/control/security"
+	"github.com/pkg/errors"
```
Comment (mjmac) [inline src/control/client/connect.go]: This import should be in its own group for 3rd-party imports. Seems to have snuck in at the end for a number of files in this PR.

File: src/control/common/system.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	m.members[member.Uuid] = member
+	m.log.Debugf("system membership: %+v\n", m.members)
+
+	return nil
+}
+
+// Remove removes member from membership.
+func (m *Membership) Remove(uuid string) error {
```
Comment (mjmac) [inline src/control/common/system.go]: Not sure if this should really return an error, because it means the caller has to check it. Does the caller really care if the to-be-removed member doesn't exist? I think we can take a cue from the delete() built-in... It does not return an error. If the map is nil or the requested member doesn't exist, the operation is a no-op.

File: src/control/common/system.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+		return errors.Errorf("member %s doesn't exist", uuid)
+	}
+
+	delete(m.members, uuid)
+
+	return nil
+}
+
+// GetMember retrieves member from membership based on UUID.
+func (m *Membership) GetMember(uuid string) SystemMember {
```
Comment (mjmac) [inline src/control/common/system.go]: ... On the other hand, I think this method should return an error if the requested member doesn't exist. Otherwise, the caller has to check for nil, and this isn't a great pattern for robust APIs.

File: src/control/common/system.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	m.Lock()
+	defer m.Unlock()
+
+	return m.members[uuid]
+}
+
+// GetMembers returns internal member structs as a sequence.
+func (m *Membership) GetMembers() []*SystemMember {
+	members := make([]*SystemMember, len(m.members))
```
Comment (mjmac) [inline src/control/common/system.go]: This should be setting the cap, not the len. Because you're using append() below vs assigning by index, the returned slice length will actually be len(m.members) * 2, with the first part of the slice full of empty elements.

File: src/control/common/system.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	for _, m := range m.members {
+		members = append(members, &m)
+	}
+
+	return members
+}
+
+// GetMembersPB converts internal member structs to protobuf equivalents.
+func (m *Membership) GetMembersPB() []*mgmtpb.SystemMember {
+	pbMembers := make([]*mgmtpb.SystemMember, len(m.members))
```
Comment (mjmac) [inline src/control/common/system.go]: Likewise.

File: src/control/server/mgmt_system.go [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+	"golang.org/x/net/context"
+
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+)
+
+// systemStop sends multicast KillRank gRPC requests to system membership list.
+func (svc *mgmtSvc) systemStop(ctx context.Context, leader *IOServerInstance) error {
+	// TODO: inhibit rebuild on pool services, parallelise and make async.
+	for _, member := range svc.members.GetMembers() {
+		svc.log.Debugf("MgmtSvc.systemStop murder member %+v (not implemented)\n",
```
Comment (mjmac) [inline src/control/server/mgmt_system.go]: While I personally find the use of "murder" in this context darkly amusing, we probably shouldn't use it in professional software. Hate to be the no-fun police...

For the record, no issue with the use of "kill", though. There is lots of precedent for that term.

File: src/control/cmd/dmg/system.go (hunk 10 above)
Comment (mjmac) [reply src/control/cmd/dmg/system.go]: FYI: https://jira.hpdd.intel.com/browse/DAOS-3407

File: src/control/common/system.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+// system running on host with the control-plane listening at "Addr".
+type SystemMember struct {
+	Addr string
+	Uuid string
+	Rank uint32
+}
+
+// Membership tracks details of system members.
+type Membership struct {
+	sync.Mutex
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/common/system.go]: okay will change

File: src/control/client/system.go [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+			rpcResp.GetStatus())
+	}
+
+	return common.MembersFromPB(rpcResp.Members), nil
+}
+
+// SystemQuery will return the list of members joined to DAOS system.
+//
+// Isolate protobuf encapsulation in client and don't expose to calling code.
+func (c *connList) SystemQuery() ([]*common.SystemMember, error) {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/client/system.go]: what about SystemMembers or SystemMemberQuery then we can align with desired syntax "dmg  system member-query"?

File: src/control/cmd/dmg/system.go (hunk 10 above)
Comment (tanabarr) [reply src/control/cmd/dmg/system.go]: makes sense

File: src/control/client/connect.go (hunk 11 above)
Comment (tanabarr) [reply src/control/client/connect.go]: that's golangci auto adding imports I'm afraid, I will keep my eye out

File: src/control/common/system.go (hunk 13 above)
Comment (tanabarr) [reply src/control/common/system.go]: agreed

File: src/control/common/system.go (hunk 14 above)
Comment (tanabarr) [reply src/control/common/system.go]: thanks, will adjust

File: src/control/server/mgmt_system.go (hunk 16 above)
Comment (tanabarr) [reply src/control/server/mgmt_system.go]: this was copying from core code, where I also found it amusing. I had meant to take it out before finalising :-)

File: src/control/client/system.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/client/system.go]: Either of those are fine with me. The latter is most precise, if a bit wordy. Is there any reason to think that you might want to keep this as a general SystemQuery with membership as one of the components of the returned value, though?

[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/client/system.go]: I was thinking about health stuff... But maybe it's best to just follow the YAGNI principle and focus on what's needed right now.

[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/client/system.go]: yes was thinking of keeping it lean, not planning too far in advance

File: src/control/cmd/dmg/system.go (hunk 10 above)
Comment (tanabarr) [reply src/control/cmd/dmg/system.go]: have adjusted to try to meet this requirement in some way to stick to the scope of the PR, the rest of the changes to respond with number of failed attempts will be dealt with in a separate PR (another subtask in the parent PR) https://jira.hpdd.intel.com/browse/DAOS-3411

File: src/control/client/connect.go (hunk 11 above)
Comment (tanabarr) [reply src/control/client/connect.go]: fixed

**Comment by tanabarr** [comment]
gatekeeper please revise commit message to take into account squashed PR#1182

File: src/control/server/mgmt_svc.go [hunk 19]
```diff
@@ -198,9 +215,27 @@ func (svc *mgmtSvc) Join(ctx context.Context, req *mgmtpb.JoinReq) (*mgmtpb.Join
...
-		svc.members.Add(common.SystemMember{
-			Addr: req.GetAddr(), Uuid: req.GetUuid(), Rank: resp.GetRank(),
-		})
+		newMember := common.SystemMember{Uuid: req.GetUuid(), Rank: resp.GetRank()}
+
+		reqAddr, err := net.ResolveTCPAddr("tcp", req.GetAddr())
+		if err != nil {
+			// leave newMember.Addr uninitialised
+			svc.log.Errorf("resolving request address: %s", err)
+			svc.members.Add(newMember)
```
Comment (mjmac) [inline src/control/server/mgmt_svc.go]: Also here... If the goal is to build up a membership set which is used to figure out who should get RPCs, adding members with unresolvable addresses seems like it will lead to problems later.

File: src/control/server/mgmt_svc.go [hunk 20]
```diff
@@ -323,3 +358,27 @@ func (svc *mgmtSvc) SmdListDevs(ctx context.Context, req *mgmtpb.SmdDevReq) (*mg
...
+// SmdListPools implements the method defined for the Management Service.
+func (svc *mgmtSvc) SmdListPools(ctx context.Context, req *mgmtpb.SmdPoolReq) (*mgmtpb.SmdPoolResp, error) {
+	mi, err := svc.harness.GetManagementInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	svc.log.Debugf("MgmtSvc.SmdListPools dispatch, req:%+v\n", *req)
+
+	svc.mutex.Lock()
```
Comment (mjmac) [inline src/control/server/mgmt_svc.go]: Bad merge? Shouldn't the locking be handled in the client now?

File: src/control/server/mgmt_svc.go (hunk 19 above)
Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: I was struggling to work out the situation, so what should happen if a join occurs but we can't resolve the address?

Comment (mjmac) [reply src/control/server/mgmt_svc.go]: It's worth looking at what could lead to a join with an unresolvable address. If it's a "can't happen" kind of situation, then I'd say we should log it at error and reject it. If there is a valid use case for allowing a join with a bad address then that's a different story, but it should be fleshed out.

My inclination would be to log/reject as a first pass and deal with handling members with bad addresses later if it comes up that we need to support that somehow. Otherwise we risk getting into weird situations without consciously choosing to go there...

Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: I should imagine that occasional address resolution failures due to temporal network issues should be an expected failure case. happy to deal with that at a later date and I think there is even a address error field that reflects a temporary failure type https://golang.org/src/net/net.go?s=17039:17075#L545 that we might be able to leverage in the future. 

For the moment I'm fine to go with log/reject as we are at an early stage.

Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: and some point we may have to deal with different types of addressing such as from proxies where we won't have the expected details in peer and will have to do some lookups

Comment (mjmac) [reply src/control/server/mgmt_svc.go]: Yep, I agree that it's probably going to happen. The question is: What should we do about it? Adding a member with an empty address seems wrong -- it's kicking the can down the road for some other part of the codebase to deal with (perhaps multiple times). Seems better to deal with it once, in one place.

Looking at the bigger picture, I notice that the peer address resolution is being done after the dRPC join call. So we may have a situation where the member has been added to the management service but we're not recording it in the control plane's membership list.

I suggest moving the address resolution step before calling dRPC join. If we can't resolve the member's address, it may indicate a bigger problem and it would probably be best to fail quickly and noisily.

**Comment by tanabarr** [comment]
management service leader
`daos_io_server:0 DAOS I/O server (v0.6.0) process 403265 started on rank 0 (out of 1) with 1 target, 1 helper XS per target, firstcore 0, host wolf-72.wolf.hpdd.intel.com.
DEBUG 15:59:57.166699 mgmt.pb.go:492: new system member: 10.8.1.71:10001/837652cc-e30a-418a-8d09-edf40f79f867/1 (total 1)
`
joining server
`DEBUG 15:59:57.163809 mgmt_drpc.go:111: I/O server instance 0 ready: uri:"ofi+psm2://2280b02:101" nctxs:4 drpcListenerSock:"/var/run/daos_server/daos_io_server_38494.sock"
DEBUG 15:59:57.163913 harness.go:232: PMIx-less mode detected
DEBUG 15:59:57.165189 mgmt_client.go:80: join(wolf-72:10001, {Uuid:837652cc-e30a-418a-8d09-edf40f79f867 Rank:4294967295 Uri:ofi+psm2://2280b02:101 Nctxs:4 Addr:0.0.0.0:10001 XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}) begin
DEBUG 15:59:57.166915 mgmt_client.go:80: join(wolf-72:10001, {Uuid:837652cc-e30a-418a-8d09-edf40f79f867 Rank:4294967295 Uri:ofi+psm2://2280b02:101 Nctxs:4 Addr:0.0.0.0:10001 XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}) end
`
